walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,129 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.collectingAndThen;
+import com.google.common.collect.ImmutableList; 
 
-public final class FindMeetingQuery {
+/**
+ * Returns a Collection of times which are available for optional attendees and mandatory attendees.
+ * If optional attendees cannot be fit in, then only returns the times which mandatory attendees can
+ * attend
+ *
+ * @param events A list of events that may need to be accounted for
+ * @param request The specifications for the meeting to be scheduled
+ * @return A Collection of available times for requested meeting
+ */
+public final class FindMeetingQuery implements Comparator<Event> {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ImmutableList<Event> eventList = events.stream().sorted(new FindMeetingQuery()).collect(collectingAndThen(toList(), ImmutableList::copyOf));
+    int mandatoryEnd = TimeRange.START_OF_DAY;
+    int optionalEnd = TimeRange.START_OF_DAY;
+    Collection<TimeRange> mandatoryAvailableTimes = new ArrayList<TimeRange>();
+    Collection<TimeRange> optionalAvailableTimes = new ArrayList<TimeRange>();
+    for (Event event: eventList) {
+      mandatoryEnd = checkIfEventConflicts(
+          mandatoryAvailableTimes, event, request, false, mandatoryEnd);
+      optionalEnd = checkIfEventConflicts(
+          optionalAvailableTimes, event, request, true, optionalEnd);
+    }
+    addFinalRange(mandatoryEnd, request, mandatoryAvailableTimes);
+    addFinalRange(optionalEnd, request, optionalAvailableTimes);
+
+    return (optionalAvailableTimes.isEmpty() && !request.getAttendees().isEmpty())
+        ? mandatoryAvailableTimes
+        : optionalAvailableTimes;
+  }
+
+
+  /**
+   * Checks if the time between the event and the last event is enough for the requested meeting
+   * time
+   *
+   * @param availableTimes A collection of times which will work for the request to be added to
+   * @param event The  event which we will check to see if there is enough time between its start
+   *     and the last meeting's end
+   * @param request The specifications for the requested meeting
+   * @param optional A boolean which determines if we are checking optional attendees as well
+   * @param start A number which represents the end of the last meeting
+   *
+   * @return The end of this event if applicable
+   */
+  public static int checkIfEventConflicts(Collection<TimeRange> availableTimes, Event event,
```

Comment:
Can this method be private?  Or is it called from some other class?

Reply:
privatized